### PR TITLE
8285794: AsyncGetCallTrace might acquire a lock via JavaThread::thread_from_jni_environment

### PR DESCRIPTION
**Pull request by parttimenerd**
Fixes the minor possibility of ASGCT acquiring a lock.

Tested using my jdk profiling tester.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8285794](https://bugs.openjdk.org/browse/JDK-8285794): AsyncGetCallTrace might acquire a lock via JavaThread::thread_from_jni_environment

### Reviewers
 * [Martin Doerr](https://openjdk.org/census#mdoerr) (@TheRealMDoerr - **Reviewer**) ⚠️ Review applies to [ebe5bc25](https://git.openjdk.org/jdk17u-dev/pull/725/files/ebe5bc25594d288f990d3415a3ab0e2cc4513597)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk17u-dev pull/725/head:pull/725` \
`$ git checkout pull/725`

Update a local copy of the PR: \
`$ git checkout pull/725` \
`$ git pull https://git.openjdk.org/jdk17u-dev pull/725/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 725`

View PR using the GUI difftool: \
`$ git pr show -t 725`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk17u-dev/pull/725.diff">https://git.openjdk.org/jdk17u-dev/pull/725.diff</a>

</details>

### CODE REVIEW

**Comment by GoeLin** [comment]
Hi, please add the label again once this is reviewed. Thanks.

**Review by TheRealMDoerr — APPROVED** [review]
LGTM.

File: src/hotspot/share/prims/forte.cpp [hunk 1]
```diff
@@ -522,12 +522,14 @@ static void forte_fill_call_trace_given_top(JavaThread* thd,
...
-  if (trace->env_id == NULL ||
-    (thread = JavaThread::thread_from_jni_environment(trace->env_id)) == NULL ||
-    thread->is_exiting()) {
+  // Can't use thread_from_jni_environment as it may also perform a VM exit check that is unsafe to
+  // do from this context.
+  Thread* raw_thread = Thread::current_or_null_safe();
+  JavaThread* thread;
 
+  if (trace->env_id == NULL || raw_thread == NULL || !raw_thread->is_Java_thread() ||
+     (thread = ((JavaThread*)raw_thread))->is_exiting()) {
```
Comment (TheRealMDoerr) [inline src/hotspot/share/prims/forte.cpp:532]: One minor diff: indentation

Comment (parttimenerd) [reply src/hotspot/share/prims/forte.cpp:532]: What is the problem with this indentation?

Comment (GoeLin) [reply src/hotspot/share/prims/forte.cpp:532]: The original change indents line 532 by four, the backport only by three.

Comment (parttimenerd) [reply src/hotspot/share/prims/forte.cpp:532]: Thanks, I'll change it accordingly.

**Pull request closed without merging**